### PR TITLE
feat(tools): add graph and graph search

### PR DESCRIPTION
**Pull request by mttrbrts**
Adds utilities to allow navigation of the dependency tree of a model as a graph. This allows for more advanced operations like "tree-shaking" (see `findConnectedGraph`), and finding unused imports (not implemented in this PR).

```js
const visitor = new ConcertoGraphVisitor();
const graph = new DirectedGraph();
modelManager.accept(visitor, { graph });

// Find the maximal subgraph starting from 'ChangeOfAddress'
const connectedGraph = graph.findConnectedGraph('org.acme.hr@1.0.0.ChangeOfAddress');

const filteredModelManager = modelManager
  .filter(declaration => 
    connectedGraph.hasVertex(declaration.getFullyQualifiedName())
  );
```

### Changes

- Adds `filter` functions to `BaseModelManager` and `ModelFile` to allow filtering declarations based on the graph.
- Adds `ConcertoGraphVisitor` and `DirectedGraph` to `concerto-tools`, and tests
- Adds new `Declaration` type to `concerto-core`. This is a new Class between `ClassDeclaration`/`ScalarDeclaration` and `Decorated`. This allows us to refactor common functions in the base class. This better matches the metamodel, will allow for better docs, better usage in TypeScript clients, and better extensibility to aggregate types in the future. We use this new type in the definition of the filter function.
- Also fixes and extends TypeScript types.

### Flags
- While refactoring `ClassDeclaration`/`ScalarDeclaration`, I noticed that `ScalarDeclaration` has several functions that should only appear on `ClassDeclaration`. I've marked them as `@deprecated`.

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [x] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [x] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: packages/concerto-core/test/introspect/declaration.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+'use strict';
+
+const Declaration = require('../../lib/introspect/declaration');
+const ModelFile = require('../../lib/introspect/modelfile');
+
+require('chai').should();
+const should = require('chai').should();
+const sinon = require('sinon');
+
+describe('Decorated', () => {
```
Comment (dselman) [inline packages/concerto-core/test/introspect/declaration.js:24]: Declaration?

File: packages/concerto-core/test/introspect/scalardeclaration.js [hunk 2]
```diff
@@ -102,11 +102,9 @@ describe('ScalarDeclaration', () => {
             const modelFiles = introspectUtils.loadModelFiles([modelFileName], modelManager);
             modelManager.addModelFiles(modelFiles);
         });
-        it('should throw an error', () => {
+        it('should return null', () => {
             const testClass = modelManager.getType('com.testing.SSN');
-            (() => {
-                testClass.getSuperType();
-            }).should.throw('Scalars do not have super types.');
+            should.equal(testClass.getSuperType(), null);
```
Comment (dselman) [inline packages/concerto-core/test/introspect/scalardeclaration.js:107]: Why return null here?

**Review by dselman — CHANGES_REQUESTED** [review]
Awesome.

File: packages/concerto-core/test/introspect/scalardeclaration.js (hunk 2 above)
Comment (mttrbrts) [reply packages/concerto-core/test/introspect/scalardeclaration.js:107]: Good question. I'm adding a test to increase coverage which reflects the current implementation.

 This is one of the now deprecated functions. It should go away in the next major release, IMO

**Review by dselman — APPROVED** [review]
Nice work!

> Note that markdown-transform uses the https://www.npmjs.com/package/dijkstrajs library. Perhaps we can refactor this (or that)?

https://github.com/accordproject/markdown-transform/blob/a6c33ba41969f7a6241a68c1aaf6c066d116a900/packages/markdown-transform/package.json#L98

**Pull request merged**